pkg/remote/remote.go

```diff
@@ -0,0 +1,320 @@
+	// be established
+	ErrRequestFailed = errors.New("can't connect to the remote host")
+	// ErrInvalidVariables is used when the variables can't be extracted from
+	// the request
+	ErrInvalidVariables = errors.New("the variables are not valid")
+	// ErrInvalidContentType is used when the response has a content-type that
+	// we deny for security reasons
+	ErrInvalidContentType = errors.New("the content-type for the response is not authorized")
+)
+
+const rawURL = "https://raw.githubusercontent.com/cozy/cozy-doctypes/master/%s/request"
+
+// Doctype is used to describe a doctype, its request for a remote doctype for example
+type Doctype struct {
+	DocID   string `json:"_id,omitempty"`
+	DocRev  string `json:"_rev,omitempty"`
+	Request string `json:"request"`
+}
+
+// ID is used to implement the couchdb.Doc interface
+func (d *Doctype) ID() string { return d.DocID }
+
+// Rev is used to implement the couchdb.Doc interface
+func (d *Doctype) Rev() string { return d.DocRev }
+
+// SetID is used to implement the couchdb.Doc interface
+func (d *Doctype) SetID(id string) { d.DocID = id }
+
+// SetRev is used to implement the couchdb.Doc interface
+func (d *Doctype) SetRev(rev string) { d.DocRev = rev }
+
+// DocType implements couchdb.Doc
+func (d *Doctype) DocType() string { return consts.Doctypes }
+
+// Clone implements couchdb.Doc
+func (d *Doctype) Clone() couchdb.Doc { cloned := *d; return &cloned }
+
+// Request is used to log in couchdb a call to a remote website
+type Request struct {
+	DocID         string `json:"_id,omitempty"`
+	DocRev        string `json:"_rev,omitempty"`
+	RemoteDoctype string `json:"doctype"`
+	Verb          string `json:"verb"`
+	URL           string `json:"url"`
+	ResponseCode  int    `json:"response_code"`
+	ContentType   string `json:"content_type"`
+}
+
+// ID is used to implement the couchdb.Doc interface
+func (r *Request) ID() string { return r.DocID }
+
+// Rev is used to implement the couchdb.Doc interface
+func (r *Request) Rev() string { return r.DocRev }
+
+// SetID is used to implement the couchdb.Doc interface
+func (r *Request) SetID(id string) { r.DocID = id }
+
+// SetRev is used to implement the couchdb.Doc interface
+func (r *Request) SetRev(rev string) { r.DocRev = rev }
+
+// DocType implements couchdb.Doc
+func (r *Request) DocType() string { return consts.RemoteRequests }
+
+// Clone implements couchdb.Doc
+func (r *Request) Clone() couchdb.Doc { cloned := *r; return &cloned }
+
+// Remote is the struct used to call a remote website for a doctype
+type Remote struct {
+	Verb    string
+	URL     *url.URL
+	Headers map[string]string
+	Body    string
+}
+
+var log = logger.WithNamespace("remote")
+
+// ParseRawRequest takes a string and parse it as a remote struct.
+// First line is verb and URL.
+// Then, we have the headers.
+// And for a POST, we have a blank line, and then the body.
+func ParseRawRequest(doctype, raw string) (*Remote, error) {
+	lines := strings.Split(raw, "\n")
```

Comment:
It's not the same format. Ours is:

```
GET https://www.example.org/foo
...
```

For `http.ReadRequest`, it should be:

```
GET /foo HTTP/1.0
Host: www.example.org
...
```

And I also think it doesn't play well with template markers.